### PR TITLE
docs: add hardware deployment guide for OpenVINO B70 and RTX 4070

### PR DESCRIPTION
**Pull request by daviburg**
## Summary

Adds comprehensive hardware deployment documentation for running semantic extraction on local GPU servers, based on measured performance from the OpenVINO B70 investigation.

### Changes

**docs/usage.md** — New "Using OpenVINO (Intel Arc / Xeon GPUs)" section:
- Tested hardware table (Arc Pro B70 31GB, RTX 4070 12GB)
- OpenVINO server setup instructions and client `config/llm.json` example
- Server batching behavior with measured throughput at batch sizes 1–8
- Concurrency rules explaining how `parallel_workers` × external workers interact
- RTX 4070 config using llama-server with `-np 4` parallel slots (not Ollama)
- `retry_failed_turns.py` usage for re-extracting failed turns

Also updates the "Tested Models" table to include `qwen3-8b-int4-ov` and adds the 31GB VRAM tier to the quick reference.

**tools/retry_failed_turns.py** — New tool for retrying failed extraction turns:
- Reads `extraction-log.jsonl` to identify turns where discovery failed
- Re-extracts them with proper concurrency control (defaults to `--workers 1`)
- Idempotent — skips turns that have since succeeded
- Includes detailed concurrency documentation in the docstring

### CODE REVIEW

File: tools/retry_failed_turns.py [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+        # Format: turn-NNN-speaker.md
+        parts = basename.replace(".md", "").split("-")
+        turn_id = f"turn-{parts[1]}"
+        speaker = parts[2] if len(parts) > 2 else "dm"
+        with open(tf, encoding="utf-8") as f:
+            text = f.read()
+        # Strip the header line (## turn-NNN [speaker])
+        lines = text.strip().split("\n")
+        if lines and lines[0].startswith("##"):
+            lines = lines[1:]
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/retry_failed_turns.py]: Addressed in a1e89fc — header stripping now handles both H1 (`# ...`) and H2 (`## ...`) headers, and strips the following blank line if present.

File: tools/retry_failed_turns.py [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+        existing_signals = {
+            (t.get("turn_id", ""), t.get("signal", ""))
+            for t in base_timeline
+        }
+        for entry in result_timeline:
+            key = (entry.get("turn_id", ""), entry.get("signal", ""))
+            if key not in existing_signals:
+                base_timeline.append(entry)
+                existing_signals.add(key)
+
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/retry_failed_turns.py]: Addressed in a1e89fc — timeline merging now uses `merge_temporal_signals()` from `temporal_extraction.py` which has proper dedup key logic (source_turn/type/season/raw_text).

File: tools/retry_failed_turns.py [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+) -> tuple[dict, list, list]:
+    """Merge entity additions from parallel extractions into base catalogs.
+
+    Each result contains a full catalog snapshot. We diff against the base
+    to find new/updated entities and merge them sequentially.
+    """
+    base_entity_ids: dict[str, set[str]] = {}
+    for key, entities in base_catalogs.items():
+        base_entity_ids[key] = {e.get("id", "") for e in entities}
+
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/retry_failed_turns.py]: Addressed in a1e89fc — removed the unused `base_entity_ids` variable.

File: tools/retry_failed_turns.py [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+                continue
+            try:
+                entry = json.loads(line)
+            except json.JSONDecodeError:
+                continue
+            turn_id = entry.get("turn_id", "")
+            if entry.get("discovery_ok"):
+                succeeded.add(turn_id)
+                failed.discard(turn_id)
+            else:
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/retry_failed_turns.py]: Addressed in a1e89fc — `load_failed_turns()` now uses `is True` / `is False` identity checks instead of truthiness, and skips entries with empty turn_id.

File: tools/retry_failed_turns.py [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+from temporal_extraction import load_timeline, save_timeline
+
+
+def load_failed_turns(extraction_log_path: str) -> list[str]:
+    """Read extraction log and return turn IDs where discovery failed.
+
+    A turn is considered failed if its most recent log entry has
+    discovery_ok=False. If a turn has both a failure and a later success,
+    the success wins (it's already been retried successfully).
+    """
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/retry_failed_turns.py:70]: Addressed in a1e89fc — added 14 unit tests in `tests/test_retry_failed_turns.py` covering `load_failed_turns()` (6 tests: failure detection, success-clears-failure, missing discovery_ok, empty turn_id, missing file, malformed JSONL), `load_turn_dicts()` (4 tests: H1/H2 header stripping, blank line removal, no-header preservation), and `merge_parallel_results()` (4 tests: skip-failed, entity merge, event dedup, timeline merge with proper dedup).

File: docs/usage.md [hunk 6]
```diff
@@ -174,6 +176,183 @@ restarts.
...
+# On the inference server (e.g. Ubuntu + Intel Arc Pro B70)
+pip install openvino openvino-genai fastapi uvicorn
+
+# Export model to OpenVINO IR format (one-time)
+optimum-cli export openvino --model Qwen/Qwen3-8B --weight-format int4_sym \
+    --trust-remote-code ./models/qwen3-8b-int4-ov
+
+# Start the server
+python serve_openvino.py --model ./models/qwen3-8b-int4-ov --port 8000
+```
```
[the remark this answers is not in the document]
Comment (daviburg) [reply docs/usage.md:209]: Addressed in a1e89fc — replaced `python serve_openvino.py` with the correct command `python -m openvino_genai.server` which is the actual OpenVINO GenAI server entry point.

File: docs/usage.md [hunk 7]
```diff
@@ -174,6 +176,183 @@ restarts.
...
+a FastAPI wrapper:
+
+```bash
+# On the inference server (e.g. Ubuntu + Intel Arc Pro B70)
+pip install openvino openvino-genai fastapi uvicorn
+
+# Export model to OpenVINO IR format (one-time)
+optimum-cli export openvino --model Qwen/Qwen3-8B --weight-format int4_sym \
+    --trust-remote-code ./models/qwen3-8b-int4-ov
+
```
[the remark this answers is not in the document]
Comment (daviburg) [reply docs/usage.md:205]: Addressed in a1e89fc — added `optimum[openvino]` to the pip install prerequisites so `optimum-cli export openvino` is runnable as documented.

File: tools/retry_failed_turns.py [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+                    "timestamp": datetime.now(timezone.utc).isoformat(timespec="seconds"),
+                    "discovery_ok": False,
+                    "discovery_error": f"QuotaExhaustedError: {e}",
+                    "detail_ok": False, "detail_error": None,
+                    "pc_ok": False, "pc_error": None,
+                    "relationships_ok": False, "relationships_error": None,
+                    "events_ok": False, "events_error": None,
+                    "new_entities": 0, "new_events": 0, "elapsed_ms": 0,
+                })
+            except Exception as e:
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/retry_failed_turns.py:346]: Addressed in a1e89fc — on `QuotaExhaustedError`, the tool now calls `pending.cancel()` on all submitted futures and breaks out of the retry loop to stop early, matching the main pipeline behavior.

File: tests/test_retry_failed_turns.py [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+
+Covers:
+- load_failed_turns() extraction log parsing
+- load_turn_dicts() transcript header stripping
+- merge_parallel_results() entity/event/timeline merging
+"""
+import json
+import os
+import sys
+import tempfile
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tests/test_retry_failed_turns.py]: Valid — `tempfile` was imported but the tests use `tmp_path` (pytest fixture) instead. Will remove the unused import.

File: tools/retry_failed_turns.py [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+            turn_id = entry.get("turn_id", "")
+            if not turn_id:
+                continue
+            if entry.get("discovery_ok") is True:
+                succeeded.add(turn_id)
+                failed.discard(turn_id)
+            elif entry.get("discovery_ok") is False:
+                if turn_id not in succeeded:
+                    failed.add(turn_id)
+
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/retry_failed_turns.py:95]: Valid — the current logic permanently masks later failures once a turn succeeds. The fix should `discard` from `succeeded` when a later `discovery_ok=False` is seen, so the last entry truly wins. Will fix in a follow-up commit.

File: tools/retry_failed_turns.py [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+        # Merge new events (deduplicate by event ID)
+        existing_event_ids = {e.get("id", "") for e in base_events}
+        for event in result_events:
+            if event.get("id", "") not in existing_event_ids:
+                base_events.append(event)
+                existing_event_ids.add(event.get("id", ""))
+
+        # Merge timeline entries using canonical dedup logic
+        merge_temporal_signals(base_timeline, result_timeline)
+
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/retry_failed_turns.py:190]: Valid concern for `--workers > 1`. In practice, `retry_failed_turns` processes turns sequentially (each worker handles one turn at a time from the failed set, and results are merged after all complete). The per-turn delta approach would be more correct for true parallel execution. Acknowledged as a known limitation for now.

File: tests/test_retry_failed_turns.py [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+        assert result == []
+
+    def test_missing_discovery_ok_not_treated_as_failure(self, tmp_path):
+        """Records without discovery_ok field are skipped (not treated as failure)."""
+        log_path = str(tmp_path / "extraction-log.jsonl")
+        with open(log_path, "w") as f:
+            f.write(json.dumps({"turn_id": "turn-010", "some_other_field": True}) + "\n")
+
+        result = load_failed_turns(log_path)
+        assert result == []
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tests/test_retry_failed_turns.py:59]: The existing `test_success_after_failure_clears_failed` covers failure→success ordering. The success→failure case is not tested because the current logic has the masking bug noted in the other thread. Will add this test alongside the fix.

File: tests/test_retry_failed_turns.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tests/test_retry_failed_turns.py]: Fixed in 2ceaea8 — removed unused `tempfile` import.

File: tools/retry_failed_turns.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/retry_failed_turns.py:95]: Filed as #292 for follow-up — will fix the last-entry-wins logic and add tests.

File: tools/retry_failed_turns.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/retry_failed_turns.py:190]: Filed as #292 — acknowledged as known limitation, tracked for future improvement.

File: tests/test_retry_failed_turns.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tests/test_retry_failed_turns.py:59]: Filed as #292 — will add success→failure test alongside the logic fix.
